keps/sig-api-machinery/2896-openapi-v3/README.md

```diff
@@ -0,0 +1,563 @@
+
+#### Proto Models & ETags
+
+Kubernetes publishes the openapi schema in both JSON and Protobuf. v3 schemas will include the same mechanisms as v2, publishing a protobuf model of the schema, as well as the corresponding etags for caching. In order to publish the schema based on groups, the CRDs will be grouped by group and published only at the endpoint for their specific group.
+
+#### Aggregator
+
+The aggregator has a mapping of all the APIServices and refreshes the aggregated spec if a single spec or CRD changes. Since the underlying api services will publish resources based on group, the aggregator must be able to discover all group endpoints for each APIService. This will increase the number of requests on bootstrap and subsequently the number of internal requests on a schema change. However, since aggregation is done by group, only a small subset of groups will be invalidated and reaggregated.
+
+### OpenAPI
+
+Swagger 2.0 and OpenAPI 3.0 have a couple of incompatible changes that are not currently supported by kube-openapi. Definitions are moved to components and the schema for Paths is slightly modified. A new struct and builder will be created in kube-openapi to represent the v3 schema to be imported by the necessary consumers.
+
+CRD structural schemas will be published as is without any value stripping for v3. Built-in types will generate the v3 OpenAPI definition, and either keep all fields or strip incompatible fields when building the swagger spec for v3 and v2 respectively.
+
+The OpenAPI handler will have an additional map to handle the publishing of v3.
+
+```go
+// Group -> OpenAPIv3 mapping
+var v3Schemas map[string]*OpenAPIv3
+
+type OpenAPIv3 struct {
+  rwMutex sync.RWMutex
+  lastModified time.Time
+
+  spec3Bytes []byte
+  spec3Pb []byte
+  spec3PbGz []byte
+
+  spec3BytesETag string
+  spec3PbETag string
+  spec3PbGzETag string
+}
+```
+
+### Test Plan
+
+<!--
+**Note:** *Not required until targeted at a release.*
+
+Consider the following in developing a test plan for this enhancement:
+- Will there be e2e and integration tests, in addition to unit tests?
+- How will it be tested in isolation vs with other components?
+
+No need to outline all of the test cases, just the general strategy. Anything
+that would count as tricky in the implementation, and anything particularly
+challenging to test, should be called out.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
+when drafting this test plan.
+
+[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
+-->
+
+For alpha, unit tests will be included to ensure that v3 schemas are properly generated and published.
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Initial e2e tests completed and enabled
```

Comment:
I want to see a beta phase explicitly planned where we are still able to fix wrong specs (especially of native types).